[Screen Time Refactoring] Do not show shield and perform donations for offscreen web views

jesxilin opened the pull request:
#### 861cf0f5c389289bcc249a923b2d0a476f8f42b3
<pre>
[Screen Time Refactoring] Do not show shield and perform donations for offscreen web views
<a href="https://bugs.webkit.org/show_bug.cgi?id=287445">https://bugs.webkit.org/show_bug.cgi?id=287445</a>
<a href="https://rdar.apple.com/144572971">rdar://144572971</a>

Reviewed by Aditya Keerthi.

Added function and logic _showScreenTimeShieldForWindow which
checks for two cases:
- if we are showing the system Screen Time blocking view,
    hide or show the view depending on if the webview is on/offscreen
- if we are showing a blurred blocking view instead of the system
    blocking view _and_ the view is currently not nil,
    hide or show the view depending on if the webview is on/offscreen

Note that in _showScreenTimeShieldForWindow, we check if we
_showsSystemScreenTimeBlockingView in both cases instead of assuming
a standard if-else case because the blurred blocking view is added and
removed _on demand_ while the system Screen Time blocking view is only
added once and removed once.

This is because the system Screen Time blocking view will
be painted/showed or not depending on if the URLIsBLocked.

Created three new API tests:
- offscreenSystemScreenTimeBlockingView
- offscreenBlurredScreenTimeBlockingView
- doNotDonateURLsInOffscreen, which tests if we suppress usage recording
    when the webview is on/offscreen

To not track URLs while offscreen, we need to set suppress usage recording
if either in private browsing or not in window
(which includes being occluded on macOS but not on iOS).

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _uninstallScreenTimeWebpageController]):
(-[WKWebView _updateScreenTimeShieldVisibilityForWindow]):
(-[WKWebView _screenTimeBlurredSnapshot]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::viewIsBecomingVisible):
(WebKit::PageClientImplCocoa::viewIsBecomingInvisible):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::viewIsBecomingInvisible):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::viewIsBecomingInvisible):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::viewIsBecomingVisible):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(TEST(ScreenTime, OffscreenSystemScreenTimeBlockingView)):
(TEST(ScreenTime, OffscreenBlurredScreenTimeBlockingView)):
(TEST(ScreenTime, DoNotDonateURLsInOffscreenWebView)):

Canonical link: <a href="https://commits.webkit.org/290486@main">https://commits.webkit.org/290486@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cc0cf454e1daacc6e95540035ba141af502ebe05

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90044 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9573 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44945 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95044 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40817 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92096 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9960 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17862 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69313 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26917 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93045 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7619 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81680 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49675 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7347 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36037 "Found 8 new test failures: fast/html/process-end-tag-for-inbody-crash.html fonts/monospace.html fonts/sans-serif.html fonts/serif.html fullscreen/fullscreen-cancel-after-request-crash.html media/modern-media-controls/fullscreen-button/fullscreen-button.html media/modern-media-controls/invalid-placard/invalid-placard-constrained-metrics.html media/modern-media-controls/placard-support/placard-support-airplay-fullscreen.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39951 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77678 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37107 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96870 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17232 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12638 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78314 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17488 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77504 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77520 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21974 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20557 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10426 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14180 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17242 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22567 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16983 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20435 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18766 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -448,10 +448,42 @@ - (void)_uninstallScreenTimeWebpageController
+    if (_screenTimeBlurredSnapshot) {
+        [_screenTimeBlurredSnapshot removeFromSuperview];
+        _screenTimeBlurredSnapshot = nil;
+    }
```
Optional suggestion, you could make this more concise. If `_screenTimeBlurredSnapshot` was already `nil`, then it's a no-op.

```objc
[std::exchange(_screenTimeBlurredSnapshot, nil) removeFromSuperview];
```

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -448,10 +448,42 @@ - (void)_uninstallScreenTimeWebpageController
+
+- (void)_showScreenTimeShieldForWindow
+{
+    bool showShield = [self window] ? true : false;
```
```suggestion
    bool showShield = [self window];
```

pxlcoder replied:
```suggestion
    BOOL showShield = !!self.window;
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -448,10 +448,42 @@ - (void)_uninstallScreenTimeWebpageController
     _screenTimeWebpageController = nil;
 }
+
+- (void)_showScreenTimeShieldForWindow
```
```suggestion
- (void)_updateScreenTimeShieldVisibilityForWindow
```

It both shows and hides.

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -448,10 +448,42 @@ - (void)_uninstallScreenTimeWebpageController
+
+    // To not track URLs while offscreen, we need to set suppress usage recording if either in private browsing or not in window (which includes being occluded on macOS).
+#if PLATFORM(MAC)
+    BOOL isNowVisible = [self window] && (([self window].occlusionState & NSWindowOcclusionStateVisible) == NSWindowOcclusionStateVisible);
```
Nit: Why using `BOOL` for this?

pxlcoder replied:
I think we should prefer `BOOL` in pure Obj-C code...

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -448,10 +448,42 @@ - (void)_uninstallScreenTimeWebpageController
+            [[_screenTimeWebpageController view] setHidden:YES];
+    }
+
+    // To not track URLs while offscreen, we need to set suppress usage recording if either in private browsing or not in window (which includes being occluded on macOS).
```
This comment isn't necessary, it's explained by the code. But it could be in the commit message.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -448,10 +448,42 @@ - (void)_uninstallScreenTimeWebpageController
+    bool showShield = [self window] ? true : false;
+
+    if (showShield) {
+        if (![_configuration _showsSystemScreenTimeBlockingView] && _screenTimeBlurredSnapshot)
```
You could store `[_configuration _showsSystemScreenTimeBlockingView]` in a local variable.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -448,10 +448,42 @@ - (void)_uninstallScreenTimeWebpageController
+
+    // To not track URLs while offscreen, we need to set suppress usage recording if either in private browsing or not in window (which includes being occluded on macOS).
+#if PLATFORM(MAC)
+    BOOL isNowVisible = [self window] && (([self window].occlusionState & NSWindowOcclusionStateVisible) == NSWindowOcclusionStateVisible);
```
```suggestion
    BOOL viewIsVisible = self.window && ((self.window.occlusionState & NSWindowOcclusionStateVisible) == NSWindowOcclusionStateVisible);
```

jesxilin replied:
Why would we do `BOOL showShield = !![self window];` and then here do `self.window`?

pxlcoder replied:
(the ask above was to do `BOOL showShield = !!self.window;`, not `[self window]`)

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -448,10 +448,42 @@ - (void)_uninstallScreenTimeWebpageController
+
+    [_screenTimeWebpageController setSuppressUsageRecording:(![_configuration websiteDataStore].isPersistent || !isNowVisible)];
+#else
+    [_screenTimeWebpageController setSuppressUsageRecording:(![_configuration websiteDataStore].isPersistent || ![self window])];
```
```suggestion
    [_screenTimeWebpageController setSuppressUsageRecording:(![_configuration websiteDataStore].isPersistent || !self.window)];
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2038,6 +2070,19 @@ - (STWebpageController *)_screenTimeWebpageController
+{
+    return _screenTimeBlurredSnapshot.get();
+}
+#endif
```
```suggestion
#if PLATFORM(MAC)
- (NSVisualEffectView *)_screenTimeBlurredSnapshot
#else
- (UIVisualEffectView *)_screenTimeBlurredSnapshot
#endif
{
    return _screenTimeBlurredSnapshot.get();
}
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -513,6 +514,11 @@ struct PerWebProcessState {
 #if ENABLE(SCREEN_TIME)
 - (STWebpageController *)_screenTimeWebpageController;
+#if PLATFORM(MAC)
+- (NSVisualEffectView *) _screenTimeBlurredSnapshot;
```
```suggestion
- (NSVisualEffectView *)_screenTimeBlurredSnapshot;
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -434,4 +439,90 @@ - (void)userContentController:(WKUserContentController *)userContentController d
 
 #endif
+
+TEST(ScreenTime, offscreenSystemScreenTimeBlockingView)
```
```suggestion
TEST(ScreenTime, OffscreenSystemScreenTimeBlockingView)
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -434,4 +439,90 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+    EXPECT_FALSE([[webView _screenTimeBlurredSnapshot] isHidden]);
+}
+
+TEST(ScreenTime, doNotDonateURLsInOffscreen)
```
```suggestion
TEST(ScreenTime, DoNotDonateURLsInOffscreenWebView)
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -434,4 +439,90 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+    EXPECT_FALSE([[[webView _screenTimeWebpageController] view] isHidden]);
+}
+
+TEST(ScreenTime, offscreenBlurredScreenTimeBlockingView)
```
```suggestion
TEST(ScreenTime, OffscreenBlurredScreenTimeBlockingView)
```

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
```diff
@@ -368,7 +368,26 @@
+
+#endif
+
+void PageClientImplCocoa::viewIsBecomingVisible()
```
`class PageClientImpl final : public PageClientImplCocoa` overrides this again on iOS, so this doesn't seem quite right, as none of this code would run on iOS?

You need to call `PageClientImplCocoa::viewIsBecomingVisible()` in `PageClientImpl::viewIsBecomingVisible`.

pxlcoder replied:
Or alternatively, delete that override and move the logic into here:

```
#if ENABLE(PAGE_LOAD_OBSERVER)
    if (RetainPtr webView = this->webView())
        [webView _updatePageLoadObserverState];
#endif
```

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -434,4 +439,90 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+}
+
 #endif
+
```
```suggestion
```

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
```diff
@@ -157,6 +158,8 @@ class PageClientImplCocoa : public PageClient {
     void setFullScreenClientForTesting(std::unique_ptr<WebFullScreenManagerProxyClient>&&) final;
 #endif
+    void viewIsBecomingVisible() override;
+    void viewIsBecomingInvisible() override;
```
```suggestion
    void viewIsBecomingInvisible() override;

```

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14668,6 +14671,7 @@ void WebPageProxy::webViewDidMoveToWindow()
         if (m_drawingArea)
             m_drawingArea->windowKindDidChange();
     }
+    pageClient->showScreenTimeShieldForWindow();
```
```suggestion

#if ENABLE(SCREEN_TIME)
    pageClient->showScreenTimeShieldForWindow();
#endif
```

pxlcoder commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
```diff
@@ -167,9 +167,14 @@ - (BOOL)_isHostedInAnotherProcess;
         [webView _updatePageLoadObserverState];
+#if ENABLE(SCREEN_TIME)
+        [webView _updateScreenTimeShieldVisibilityForWindow];
 #endif
```
If you want to move the logic into here, rather than the other way around, you can just `PageClientImplCocoa::viewIsBecomingVisible()` at the top of this method.

jesxilin replied:
Okay, that sounds better, I will change it

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -448,10 +448,40 @@ - (void)_uninstallScreenTimeWebpageController
+    [_screenTimeWebpageController setSuppressUsageRecording:(![_configuration websiteDataStore].isPersistent || !isNowVisible)];
+#else
+    [_screenTimeWebpageController setSuppressUsageRecording:(![_configuration websiteDataStore].isPersistent || !self.window)];
+#endif
```
Reducing duplicate logic:

```suggestion
    BOOL viewIsInWindow = !!self.window;

    BOOL showsSystemScreenTimeBlockingView = [_configuration _showsSystemScreenTimeBlockingView];

    if (viewIsInWindow) {
        if (!showsSystemScreenTimeBlockingView && _screenTimeBlurredSnapshot)
            [_screenTimeBlurredSnapshot setHidden:NO];
        else if (showsSystemScreenTimeBlockingView)
            [[_screenTimeWebpageController view] setHidden:NO];
    } else {
        if (_screenTimeBlurredSnapshot)
            [_screenTimeBlurredSnapshot setHidden:YES];
        else if (showsSystemScreenTimeBlockingView)
            [[_screenTimeWebpageController view] setHidden:YES];
    }

    BOOL viewIsVisible = viewIsInWindow;
#if PLATFORM(MAC)
    viewIsVisible &= ((self.window.occlusionState & NSWindowOcclusionStateVisible) == NSWindowOcclusionStateVisible);
#endif
    [_screenTimeWebpageController setSuppressUsageRecording:(![_configuration websiteDataStore].isPersistent || !viewIsVisible)];
```

jesxilin replied:
Ooh, thank you!

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14668,6 +14671,9 @@ void WebPageProxy::webViewDidMoveToWindow()
     }
+#if ENABLE(SCREEN_TIME)
+    pageClient->updateScreenTimeShieldVisibilityForWindow();
+#endif
```
Do we really need this call, or does `viewIsBecomingVisible` / `viewIsBecomingInvisible` get called when adding/removing from the window too? I would expect that it does, and that this call redundant, but I might be wrong.

jesxilin replied:
I notice without this call, the API test for `OffscreenBlurredScreenTimeBlockingView` and `OffscreenSystemScreenTimeBlockingView` will not pass. I think this because we remove from window, which is neither `viewIsBecomingVisible` or `viewIsBecomingInvisible` but `webViewDidMoveToWindow`. So in that case, we do need the call.

jesxilin replied:
Follow up: Activity states are asynchronous so by waiting in the API test, seems it actually passes and viewIsBecomingVisible/viewIsBecomingInvisible do get called. Will test again and remove!

jesxilin committed:
[Screen Time Refactoring] Do not show shield and perform donations for offscreen web views
https://bugs.webkit.org/show_bug.cgi?id=287445
rdar://144572971

Reviewed by Aditya Keerthi.

Added function and logic _showScreenTimeShieldForWindow which
checks for two cases:
- if we are showing the system Screen Time blocking view,
    hide or show the view depending on if the webview is on/offscreen
- if we are showing a blurred blocking view instead of the system
    blocking view _and_ the view is currently not nil,
    hide or show the view depending on if the webview is on/offscreen

Note that in _showScreenTimeShieldForWindow, we check if we
_showsSystemScreenTimeBlockingView in both cases instead of assuming
a standard if-else case because the blurred blocking view is added and
removed _on demand_ while the system Screen Time blocking view is only
added once and removed once.

This is because the system Screen Time blocking view will
be painted/showed or not depending on if the URLIsBLocked.

Created three new API tests:
- offscreenSystemScreenTimeBlockingView
- offscreenBlurredScreenTimeBlockingView
- doNotDonateURLsInOffscreen, which tests if we suppress usage recording
    when the webview is on/offscreen

To not track URLs while offscreen, we need to set suppress usage recording
if either in private browsing or not in window
(which includes being occluded on macOS but not on iOS).

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _uninstallScreenTimeWebpageController]):
(-[WKWebView _updateScreenTimeShieldVisibilityForWindow]):
(-[WKWebView _screenTimeBlurredSnapshot]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::viewIsBecomingVisible):
(WebKit::PageClientImplCocoa::viewIsBecomingInvisible):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::viewIsBecomingInvisible):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::viewIsBecomingInvisible):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::viewIsBecomingVisible):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(TEST(ScreenTime, OffscreenSystemScreenTimeBlockingView)):
(TEST(ScreenTime, OffscreenBlurredScreenTimeBlockingView)):
(TEST(ScreenTime, DoNotDonateURLsInOffscreenWebView)):

Canonical link: https://commits.webkit.org/290486@main
